File: backend/src/LibraryConnect.Application/Features/Digital/DigitalModels.cs
using LibraryConnect.Application.Common.Models;
using LibraryConnect.Domain.Enums;

namespace LibraryConnect.Application.Features.Digital;

// ---------------------------------------------------------------------------------------------
// Phân hệ V — Tài liệu số. Các kiểu dữ liệu dùng chung cho cả màn hình quản trị lẫn nhóm endpoint
// dành cho bạn đọc.
// ---------------------------------------------------------------------------------------------

/// <summary>Một nút trên cây bộ sưu tập (V.1).</summary>
public record DigitalCollectionDto(
    Guid Id,
    string Code,
    string Name,
    string? NameEn,
    Guid? ParentId,
    string? ParentName,
    string? Description,
    DigitalAccessLevel DefaultAccessLevel,
    int SortOrder,
    bool IsActive,
    int DocumentCount,
    IReadOnlyList<DigitalCollectionDto> Children);

/// <summary>Một dòng trên danh sách tài liệu số.</summary>
public record DigitalDocumentRowDto(
    Guid Id,
    string Title,
    string FileName,
    string MimeType,
    long FileSize,
    int? PageCount,
    Guid? CollectionId,
    string? CollectionName,
    Guid? BibId,
    string? BibTitle,
    DigitalAccessLevel AccessLevel,
    bool AllowDownload,
    bool AllowPrint,
    bool WatermarkEnabled,
    int PreviewPages,
    bool HasThumbnail,
    bool HasText,
    bool OcrProcessed,
    int ViewCount,
    int DownloadCount,
    string? UploadByName,
    DateTimeOffset UploadAt,
    // Đoạn văn bản chứa từ khóa, chỉ có khi tìm kiếm toàn văn.
    string? Snippet);

/// <summary>Chi tiết một tài liệu số kèm các bản dẫn xuất và quyền của người đang xem.</summary>
public record DigitalDocumentDetailDto(
    DigitalDocumentRowDto Document,
    string? Description,
    string? ChecksumSha256,
    IReadOnlyList<DigitalDocumentFileDto> Files,
    DigitalPermissionDto Permission);

public record DigitalDocumentFileDto(
    Guid Id, DigitalFileType Type, string Path, long Size, string? MimeType, int? PageNumber);

/// <summary>
/// Người đang gọi được làm gì với tài liệu này.
///
/// Máy chủ quyết định và trả xuống, chứ không để giao diện tự suy từ mức truy cập — mọi endpoint
/// đọc và tải đều kiểm lại đúng bộ quy tắc này.
/// </summary>
public record DigitalPermissionDto(
    bool CanRead,
    bool CanDownload,
    bool CanPrint,
    // Số trang được xem; null nghĩa là xem hết.
    int? ReadablePages,
    bool NeedsRequest,
    // Trạng thái yêu cầu bạn đọc đã gửi, nếu có.
    AccessRequestStatus? RequestStatus,
    DateTimeOffset? AccessExpireAt,
    string Reason);

/// <summary>Bộ lọc danh sách tài liệu số.</summary>
public class DigitalDocumentFilter
{
    public Guid? CollectionId { get; set; }
    /// <summary>Lấy cả tài liệu của các bộ sưu tập con.</summary>
    public bool IncludeDescendants { get; set; } = true;
    public Guid? BibId { get; set; }
    public DigitalAccessLevel? AccessLevel { get; set; }
    /// <summary>Nhóm định dạng: PDF, VIDEO, AUDIO, IMAGE, OFFICE, OTHER.</summary>
    public string? FormatGroup { get; set; }
    public bool? HasText { get; set; }
    /// <summary>Tìm trong nội dung tài liệu chứ không chỉ trong nhan đề.</summary>
    public bool FullText { get; set; }
    public DateOnly? UploadedFrom { get; set; }
    public DateOnly? UploadedTo { get; set; }
}

public class DigitalDocumentQueryRequest : PagedRequest
{
    public DigitalDocumentFilter Filter { get; set; } = new();
}

/// <summary>Một dòng trên danh sách yêu cầu đọc tài liệu hạn chế (V.2).</summary>
public record DigitalAccessRequestRowDto(
    Guid Id,
    Guid DocumentId,
    string DocumentTitle,
    Guid ReaderId,
    string ReaderName,
    string ReaderCardNumber,
    string? ReaderTypeName,
    string? FacultyName,
    DateTimeOffset RequestDate,
    string? Reason,
    AccessRequestStatus Status,
    string? ApprovedByName,
    DateTimeOffset? ApprovedAt,
    DateTimeOffset? ExpireAt,
    string? RejectReason,
    int? MaxViews,
    int ViewCount,
    bool AllowDownload,
    // Số giờ từ lúc gửi tới lúc cán bộ xử lý — dùng cho báo cáo V.4.
    double? ProcessingHours);

/// <summary>Một dòng nhật ký truy cập tài liệu số (V.2).</summary>
public record DigitalAccessLogRowDto(
    Guid Id,
    Guid DocumentId,
    string DocumentTitle,
    Guid? ReaderId,
    string? ReaderName,
    string? ReaderCardNumber,
    string? UserName,
    DigitalAccessAction Action,
    string? Ip,
    string? Device,
    int? PageFrom,
    int? PageTo,
    int? DurationSeconds,
    DateTimeOffset OccurredAt);

/// <summary>Trạng thái một phiên tải tệp lớn theo mảnh.</summary>
public record DigitalUploadSessionDto(
    Guid Id,
    string FileName,
    long TotalSize,
    int ChunkSize,
    int TotalChunks,
    IReadOnlyList<int> ReceivedChunks,
    IReadOnlyList<int> MissingChunks,
    bool IsCompleted,
    Guid? DocumentId,
    DateTimeOffset ExpiresAt);

File: mobile/lib/shared/models/digital_models.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'digital_models.freezed.dart';
part 'digital_models.g.dart';

/// Một tài liệu số bạn đọc nhìn thấy (`/api/reader/digital`, `/reader/digital/search`).
@freezed
abstract class DigitalDocumentRow with _$DigitalDocumentRow {
  const factory DigitalDocumentRow({
    required String id,
    required String title,
    @Default('') String fileName,
    @Default('application/pdf') String mimeType,
    @Default(0) int fileSize,
    int? pageCount,
    String? collectionId,
    String? collectionName,
    String? bibId,
    String? bibTitle,
    @Default('Internal') String accessLevel,
    @Default(false) bool allowDownload,
    @Default(false) bool allowPrint,
    @Default(0) int previewPages,
    @Default(0) int viewCount,
    @Default(0) int downloadCount,
    DateTime? updatedAt,
  }) = _DigitalDocumentRow;

  const DigitalDocumentRow._();

  factory DigitalDocumentRow.fromJson(Map<String, dynamic> json) =>
      _$DigitalDocumentRowFromJson(json);

  bool get isRestricted => accessLevel == 'Restricted';
  bool get isForbidden => accessLevel == 'Forbidden';
}

/// Quyền của chính bạn đọc với một tài liệu — máy chủ tính, ứng dụng chỉ hiện.
@freezed
abstract class DigitalPermission with _$DigitalPermission {
  const factory DigitalPermission({
    @Default(false) bool canRead,
    @Default(false) bool canDownload,
    @Default(false) bool canPrint,
    int? readablePages,
    @Default(false) bool needsRequest,
    String? requestStatus,
    DateTime? accessExpireAt,
    @Default('') String reason,
  }) = _DigitalPermission;

  factory DigitalPermission.fromJson(Map<String, dynamic> json) =>
      _$DigitalPermissionFromJson(json);
}

@freezed
abstract class DigitalDocumentDetail with _$DigitalDocumentDetail {
  const factory DigitalDocumentDetail({
    required DigitalDocumentRow document,
    String? description,
    String? checksumSha256,
    @Default(DigitalPermission()) DigitalPermission permission,
  }) = _DigitalDocumentDetail;

  factory DigitalDocumentDetail.fromJson(Map<String, dynamic> json) =>
      _$DigitalDocumentDetailFromJson(json);
}

/// Phiên đọc trực tuyến (`/reader/digital/{id}/read`): số trang được xem, có chữ chìm không.
@freezed
abstract class DigitalReaderSession with _$DigitalReaderSession {
  const factory DigitalReaderSession({
    required String documentId,
    required String title,
    int? pageCount,
    int? readablePages,
    @Default(false) bool canDownload,
    @Default(false) bool canPrint,
    @Default(false) bool watermarkEnabled,
    @Default('application/pdf') String mimeType,
    @Default('') String reason,
  }) = _DigitalReaderSession;

  const DigitalReaderSession._();

  factory DigitalReaderSession.fromJson(Map<String, dynamic> json) =>
      _$DigitalReaderSessionFromJson(json);

  /// Số trang mở được: giới hạn xem thử nếu có, không thì toàn bộ.
  int get pagesToShow {
    final total = pageCount ?? 0;
    final readable = readablePages;
    if (readable == null) return total;
    return readable < total ? readable : total;
  }
}

@freezed
abstract class DigitalCollectionNode with _$DigitalCollectionNode {
  const factory DigitalCollectionNode({
    required String id,
    @Default('') String code,
    required String name,
    String? parentId,
    @Default(0) int documentCount,
    @Default([]) List<DigitalCollectionNode> children,
  }) = _DigitalCollectionNode;

  factory DigitalCollectionNode.fromJson(Map<String, dynamic> json) =>
      _$DigitalCollectionNodeFromJson(json);
}

/// Một chỗ khớp khi tìm trong văn bản (`/reader/digital/{id}/find`).
@freezed
abstract class DigitalTextHit with _$DigitalTextHit {
  const factory DigitalTextHit({
    required int page,
    @Default('') String snippet,
  }) = _DigitalTextHit;

  factory DigitalTextHit.fromJson(Map<String, dynamic> json) =>
      _$DigitalTextHitFromJson(json);
}

/// Yêu cầu truy cập tài liệu hạn chế đã gửi.
@freezed
abstract class DigitalAccessRequestRow with _$DigitalAccessRequestRow {
  const factory DigitalAccessRequestRow({
    required String id,
    required String documentId,
    @Default('') String documentTitle,
    DateTime? requestDate,
    String? reason,
    @Default('Pending') String status,
    DateTime? approvedAt,
    DateTime? expireAt,
    String? rejectReason,
    int? maxViews,
    @Default(0) int viewCount,
    @Default(false) bool allowDownload,
  }) = _DigitalAccessRequestRow;

  factory DigitalAccessRequestRow.fromJson(Map<String, dynamic> json) =>
      _$DigitalAccessRequestRowFromJson(json);
}

/// Một dòng lịch sử xem / tải của chính bạn đọc.
@freezed
abstract class DigitalAccessLogRow with _$DigitalAccessLogRow {
  const factory DigitalAccessLogRow({
    required String id,
    required String documentId,
    @Default('') String documentTitle,
    @Default('View') String action,
    int? pageFrom,
    int? pageTo,
    int? durationSeconds,
    DateTime? occurredAt,
  }) = _DigitalAccessLogRow;

  factory DigitalAccessLogRow.fromJson(Map<String, dynamic> json) =>
      _$DigitalAccessLogRowFromJson(json);
}

/// Gói đọc ngoại tuyến máy chủ cấp (`POST /reader/digital/{id}/offline-package`).
@freezed
abstract class OfflinePackage with _$OfflinePackage {
  const factory OfflinePackage({
    required String packageId,
    required String documentId,
    required String title,
    @Default('') String fileName,
    @Default('application/pdf') String mimeType,
    @Default(0) int sizeBytes,
    @Default('') String checksum,
    @Default('AES-256-CBC') String algorithm,
    required String keyBase64,
    required String ivBase64,
    required DateTime expiresAt,
    required String downloadUrl,
  }) = _OfflinePackage;

  factory OfflinePackage.fromJson(Map<String, dynamic> json) =>
      _$OfflinePackageFromJson(json);
}

/// Gói ngoại tuyến đã cấp, theo danh sách máy chủ.
@freezed
abstract class OfflinePackageRow with _$OfflinePackageRow {
  const factory OfflinePackageRow({
    required String packageId,
    required String documentId,
    required String title,
    DateTime? createdAt,
    required DateTime expiresAt,
    DateTime? downloadedAt,
    @Default(false) bool isRevoked,
    @Default(false) bool isExpired,
  }) = _OfflinePackageRow;

  factory OfflinePackageRow.fromJson(Map<String, dynamic> json) =>
      _$OfflinePackageRowFromJson(json);
}

/// Một mục trong mục lục tài liệu số (`/reader/digital/{id}/outline`): độ sâu để thụt lề, trang
/// đích để nhảy (null là tiêu đề nhóm không trỏ trang). Lớp thường, không freezed, vì còn được
/// ghi kèm gói ngoại tuyến bằng `toJson` và đọc lại khi không có mạng.
class DigitalOutlineEntry {
  const DigitalOutlineEntry({
    required this.level,
    required this.title,
    this.page,
  });

  factory DigitalOutlineEntry.fromJson(Map<String, dynamic> json) =>
      DigitalOutlineEntry(
        level: (json['level'] as num?)?.toInt() ?? 0,
        title: json['title'] as String? ?? '',
        page: (json['page'] as num?)?.toInt(),
      );

  final int level;
  final String title;
  final int? page;

  Map<String, dynamic> toJson() => {
    'level': level,
    'title': title,
    'page': page,
  };

  static List<DigitalOutlineEntry> listFromJson(Object? json) => json is List
      ? json
            .whereType<Map<String, dynamic>>()
            .map(DigitalOutlineEntry.fromJson)
            .toList(growable: false)
      : const [];
}
